File: frontend/lib/project-setup-readiness.ts
import type { CentralScheme, PlantingProject, WorkArea } from "@/lib/api";
import {
  treeRegistrationDefaultsFromProject,
  validateTreeRegistrationDefaults,
} from "@/lib/tree-registration-defaults";
import { projectSetupHref } from "@/lib/project-focused-ui";
import { isMonitoringScheme } from "@/lib/schemes";

export type SetupStepId =
  | "scheme_refs"
  | "tree_defaults"
  | "planting_standard"
  | "work_areas"
  | "ready";

export type SetupStep = {
  id: SetupStepId;
  label: string;
  complete: boolean;
  required: boolean;
  href?: string;
  description?: string;
};

export type ProjectSetupStatus = {
  steps: SetupStep[];
  setupComplete: boolean;
  canRegisterTree: boolean;
  monitoringMode: boolean;
  blockReason?: string;
};

type ReadinessInput = {
  project: PlantingProject;
  workAreas: WorkArea[];
  scheme?: CentralScheme | null;
};

function schemeRefFields(scheme?: CentralScheme | null) {
  const section = scheme?.metadata_sections?.[0] as
    | { fields?: { key: string; required?: boolean }[] }
    | undefined;
  return section?.fields ?? [];
}

export function missingSchemeRefKeys(
  project: PlantingProject,
  scheme?: CentralScheme | null,
): string[] {
  if (!project.scheme_code || !scheme) return [];
  const refs =
    (project.metadata?.scheme_refs as Record<string, unknown> | undefined) ?? {};
  const missing: string[] = [];
  for (const field of schemeRefFields(scheme)) {
    if (!field.required) continue;
    const value = refs[field.key];
    if (value == null || String(value).trim() === "") {
      missing.push(field.key);
    }
  }
  return missing;
}

export function evaluateProjectSetup({
  project,
  workAreas,
  scheme,
}: ReadinessInput): ProjectSetupStatus {
  const monitoringMode = isMonitoringScheme(project.scheme_code);
  const missingRefs = missingSchemeRefKeys(project, scheme);
  const hasSchemeRefs = project.scheme_code ? missingRefs.length === 0 : true;
  const treeDefaults = treeRegistrationDefaultsFromProject(project);
  const requiresTreeDefaults =
    !monitoringMode &&
    Boolean(project.scheme_code || project.program_code === "government_nhai");
  const missingTreeDefaults = requiresTreeDefaults
    ? validateTreeRegistrationDefaults(treeDefaults)
    : {};
  const hasTreeDefaults = Object.keys(missingTreeDefaults).length === 0;
  const hasStandard = Boolean(project.active_standard);
  const hasWorkAreas = workAreas.length > 0;
  const requiresWorkArea =
    project.compliance_mode === "strict" || project.compliance_mode === "guided";

  const steps: SetupStep[] = [];

  if (project.scheme_code) {
    steps.push({
      id: "scheme_refs",
      label: monitoringMode ? "Estate details" : "Scheme references",
      complete: hasSchemeRefs,
      required: true,
      href: projectSetupHref(project.id, 3),
      description: hasSchemeRefs
        ? monitoringMode
          ? "Estate identity saved for monitoring exports"
          : "Government IDs saved for audit exports"
        : `${missingRefs.length} required field${missingRefs.length === 1 ? "" : "s"} missing`,
    });
  }

  if (requiresTreeDefaults) {
    steps.push({
      id: "tree_defaults",
      label: "Tree registration defaults",
      complete: hasTreeDefaults,
      required: true,
      href: projectSetupHref(project.id, 3),
      description: hasTreeDefaults
        ? "Permit, site zone, and agency defaults saved"
        : `Missing: ${Object.keys(missingTreeDefaults).join(", ")}`,
    });
  }

  steps.push({
    id: "planting_standard",
    label: monitoringMode ? "Monitoring standard" : "Planting standard",
    complete: hasStandard,
    required: true,
    description: hasStandard
      ? (project.active_standard?.name ?? "Standard attached")
      : monitoringMode
        ? "No monitoring standard attached"
        : "No compliance standard attached",
  });

  steps.push({
    id: "work_areas",
    label: "Work areas on map",
    complete: hasWorkAreas,
    required: requiresWorkArea,
    href: projectSetupHref(project.id, 4),
    description: hasWorkAreas
      ? `${workAreas.length} area${workAreas.length === 1 ? "" : "s"} drawn`
      : monitoringMode
        ? "Draw at least one estate block polygon (10–500 ha recommended)"
        : "Draw at least one polygon or corridor",
  });

  const setupComplete =
    hasStandard &&
    hasSchemeRefs &&
    hasTreeDefaults &&
    (!requiresWorkArea || hasWorkAreas);

  let canRegisterTree = setupComplete;
  let blockReason: string | undefined;

  if (!hasStandard) {
    canRegisterTree = false;
    blockReason = monitoringMode
      ? "Attach a monitoring standard before optional tree registration."
      : "Attach a planting standard before registering trees.";
  } else if (!hasSchemeRefs) {
    canRegisterTree = false;
    blockReason = monitoringMode
      ? "Complete estate details in project setup."
      : "Complete scheme references in project setup.";
  } else if (!hasTreeDefaults) {
    canRegisterTree = false;
    blockReason =
      "Complete tree registration defaults (permit, site zone, agency) in project setup.";
  } else if (requiresWorkArea && !hasWorkAreas) {
    canRegisterTree = false;
    blockReason = monitoringMode
      ? "Draw an estate work area on the map before registering optional ground-truth trees."
      : "Draw a work area on the map before registering trees.";
  }

  return {
    steps,
    setupComplete,
    canRegisterTree,
    monitoringMode,
    blockReason,
  };
}

File: mobile/lib/src/project_setup_readiness.dart
/// Mirrors web `evaluateProjectSetup` for mobile tree registration gates.
library;

class ProjectSetupStep {
  const ProjectSetupStep({
    required this.id,
    required this.label,
    required this.complete,
    required this.required,
    this.description,
  });

  final String id;
  final String label;
  final bool complete;
  final bool required;
  final String? description;
}

class ProjectSetupStatus {
  const ProjectSetupStatus({
    required this.steps,
    required this.canRegisterTree,
    this.blockReason,
  });

  final List<ProjectSetupStep> steps;
  final bool canRegisterTree;
  final String? blockReason;
}

const _treeDefaultKeys = [
  ('permit_reference', 'Permit reference'),
  ('site_zone', 'Site zone'),
  ('implementing_agency', 'Implementing agency'),
  ('maintenance_responsible', 'Maintenance responsible'),
];

List<String> missingSchemeRefKeys(
  Map<String, dynamic> project,
  Map<String, dynamic>? scheme,
) {
  final schemeCode = project['scheme_code'] as String?;
  if (schemeCode == null || scheme == null) return [];
  final sections = (scheme['metadata_sections'] as List<dynamic>?) ?? [];
  if (sections.isEmpty) return [];
  final fields = (sections.first as Map<String, dynamic>?)?['fields'] as List<dynamic>? ?? [];
  final refs = Map<String, dynamic>.from(
    (project['metadata'] as Map?)?['scheme_refs'] as Map? ?? {},
  );
  final missing = <String>[];
  for (final field in fields) {
    final map = Map<String, dynamic>.from(field as Map);
    if (map['required'] != true) continue;
    final key = map['key'] as String? ?? '';
    final value = refs[key];
    if (value == null || value.toString().trim().isEmpty) {
      missing.add(key);
    }
  }
  return missing;
}

ProjectSetupStatus evaluateProjectSetup(
  Map<String, dynamic> project,
  List<dynamic> workAreas, {
  Map<String, dynamic>? scheme,
}) {
  final metadata = Map<String, dynamic>.from(project['metadata'] as Map? ?? {});
  final defaults = Map<String, dynamic>.from(metadata['tree_registration_defaults'] as Map? ?? {});
  final schemeCode = project['scheme_code'] as String?;
  final programCode = project['program_code'] as String?;
  final complianceMode = project['compliance_mode'] as String? ?? 'open';
  final requiresWorkArea = complianceMode == 'strict' || complianceMode == 'guided';

  final missingRefs = missingSchemeRefKeys(project, scheme);
  final hasSchemeRefs = schemeCode == null || missingRefs.isEmpty;
  final hasStandard = project['active_standard'] != null;
  final missingDefaults = <String>[];
  if (schemeCode != null || programCode == 'government_nhai') {
    for (final (key, label) in _treeDefaultKeys) {
      if ((defaults[key]?.toString().trim() ?? '').isEmpty) {
        missingDefaults.add(label);
      }
    }
  }
  final hasTreeDefaults = missingDefaults.isEmpty;
  final hasWorkAreas = workAreas.isNotEmpty;

  final steps = <ProjectSetupStep>[
    if (schemeCode != null)
      ProjectSetupStep(
        id: 'scheme_refs',
        label: 'Scheme references',
        complete: hasSchemeRefs,
        required: true,
        description: hasSchemeRefs
            ? 'Government IDs saved for audit exports'
            : 'Missing: ${missingRefs.join(', ')}',
      ),
    if (schemeCode != null || programCode == 'government_nhai')
      ProjectSetupStep(
        id: 'tree_defaults',
        label: 'Tree registration defaults',
        complete: hasTreeDefaults,
        required: true,
        description: hasTreeDefaults
            ? 'Permit, site zone, and agency saved'
            : 'Missing: ${missingDefaults.join(', ')}',
      ),
    ProjectSetupStep(
      id: 'planting_standard',
      label: 'Planting standard',
      complete: hasStandard,
      required: true,
      description: hasStandard ? 'Compliance standard attached' : 'No planting standard attached',
    ),
    ProjectSetupStep(
      id: 'work_areas',
      label: 'Work areas on map',
      complete: hasWorkAreas,
      required: requiresWorkArea,
      description: hasWorkAreas
          ? '${workAreas.length} area${workAreas.length == 1 ? '' : 's'} defined'
          : 'Draw at least one polygon or corridor',
    ),
  ];

  String? blockReason;
  var canRegister = hasStandard && hasTreeDefaults && hasSchemeRefs && (!requiresWorkArea || hasWorkAreas);

  if (!hasSchemeRefs) {
    canRegister = false;
    blockReason = 'Complete scheme reference fields in project setup before registering trees.';
  } else if (!hasStandard) {
    canRegister = false;
    blockReason = 'Attach a planting standard before registering trees.';
  } else if (!hasTreeDefaults) {
    canRegister = false;
    blockReason =
        'Complete tree registration defaults (permit, site zone, agency) in project setup.';
  } else if (requiresWorkArea && !hasWorkAreas) {
    canRegister = false;
    blockReason = 'Draw a work area on the map before registering trees.';
  }

  return ProjectSetupStatus(
    steps: steps,
    canRegisterTree: canRegister,
    blockReason: blockReason,
  );
}

String projectSetupWebUrl(String projectId) {
  return 'https://aranyix.tech/projects/$projectId/setup';
}

String projectCreateWebUrl() => 'https://aranyix.tech/projects/new';
